receiver/vcenterreceiver/client.go

```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
 		resultsByRef: resultsByRef,
 	}, nil
 }
+
+// VSANQueryResults contains all returned vSAN metric related data
+type VSANQueryResults struct {
+	// Contains vSAN metric data keyed by UUID string
+	MetricResultsByUUID map[string]*VSANMetricResults
+}
+
+// VSANMetricResults contains vSAN metric related data for a single resource
+type VSANMetricResults struct {
+	// Contains UUID info for related resource
+	UUID string
+	// Contains returned metric value info for all metrics
+	MetricDetails []*VSANMetricDetails
+}
+
+// VSANMetricDetails contains vSAN metric data for a single metric
+type VSANMetricDetails struct {
+	// Contains the metric label
+	MetricLabel string
+	// Contains the metric interval in seconds
+	Interval int32
+	// Contains timestamps for all metric values
+	Timestamps []*time.Time
+	// Contains all values for vSAN metric label
+	Values []int64
+}
+
+// VSANQueryType represents the type of VSAN query
+type VSANQueryType string
+
+const (
+	VSANQueryTypeHosts           VSANQueryType = "host-domclient:*"
+	VSANQueryTypeVirtualMachines VSANQueryType = "virtual-machine:*"
+)
+
+// getLabelsForQueryType returns the appropriate labels for each query type
+func (vc *vcenterClient) getLabelsForQueryType(queryType VSANQueryType) []string {
+	switch queryType {
+	case VSANQueryTypeHosts:
+		return []string{
+			"iopsRead", "iopsWrite", "throughputRead", "throughputWrite",
+			"latencyRead", "latencyWrite", "congestion", "clientCacheHitRate",
+		}
+	case VSANQueryTypeVirtualMachines:
+		return []string{
+			"iopsRead", "iopsWrite", "throughputRead", "throughputWrite",
+			"latencyRead", "latencyWrite",
+		}
+	default:
+		return []string{}
+	}
+}
+
+// VSANQuery performs a VSAN query for the specified type across multiple clusters
+func (vc *vcenterClient) VSANQuery(
+	ctx context.Context,
+	queryType VSANQueryType,
+	clusterRefs []*vt.ManagedObjectReference,
+) (*VSANQueryResults, error) {
+	allResults := VSANQueryResults{
+		MetricResultsByUUID: map[string]*VSANMetricResults{},
+	}
+
+	if vc.vsanDriver == nil {
+		return &allResults, nil
+	}
+
+	for _, clusterRef := range clusterRefs {
+		results, err := vc.vsanQueryByCluster(ctx, queryType, clusterRef)
+		if err != nil {
+			vsanErr := vc.handleVSANError(err, queryType)
+			if vsanErr != nil {
+				return nil, vsanErr
+			}
+			continue
```

Comment:
This continue is probably fine (compared to returning "nil, &allResults"). My only concern here is that we don't have a test for this situation right now. To be fair, we also don't have a test for the happy path of multiple clusters either for vSAN as well. It would be a bit of a PITA, but it might be worth modifying the scrape tests/results to have a 2nd cluster with a single resource pool/host/VM.

Really....this would normally be handled in the client tests with the simulator, but the vSAN simulator isn't working.

Reply:
I can modify the scrape tests, yea its a shame this won't be handled with the govmomi simulator, but I think this is a close second best alternative.

Reply:
@BominRahmani Soooo, maybe you should take a look at the VM/Cluster PRs that I have out. I no longer am passing in a list of clusters, and this continue isn't actually needed...

Reply:
resolve